t/unit/tasks/test_canvas.py

```diff
@@ -607,6 +607,10 @@ def test_app_when_app_in_task(self):
         x = chord([t1], body=t2)
         assert x.app is t2._app
 
+    def test_app_when_header_is_empty(self):
+        x = chord([], self.add.s(4, 4))
+        assert x.app is self.add.app
```

Comment:
Indeed the test you added seems to verify the failure case (tried it with `master`).